File: rp_engine/models/rp.py
"""Pydantic models for RP management and guidelines."""

from __future__ import annotations

from pydantic import BaseModel

from rp_engine.models.enums import (
    NarrativeVoice,
    PovMode,
    ResponseLength,
    ScenePacing,
    Tense,
)


class ChunkingConfig(BaseModel):
    strategy: str = "fixed"
    chunk_size: int = 1000
    chunk_overlap: int = 200


class ChunkingUpdate(BaseModel):
    strategy: str | None = None
    chunk_size: int | None = None
    chunk_overlap: int | None = None


class RechunkResponse(BaseModel):
    status: str
    total_exchanges: int = 0
    embedded: int = 0
    skipped: int = 0
    failed: int = 0


class RPCreate(BaseModel):
    rp_name: str
    pov_mode: PovMode = "single"
    dual_characters: list[str] = []
    tone: str = ""
    scene_pacing: ScenePacing = "moderate"
    narrative_voice: NarrativeVoice = "third"
    tense: Tense = "past"
    response_length: ResponseLength = "medium"


class RPResponse(BaseModel):
    rp_folder: str
    created_files: list[str] = []


class RPInfo(BaseModel):
    rp_folder: str
    has_story_cards: bool
    card_count: int
    has_guidelines: bool
    has_avatar: bool = False
    branches: list[str] = []


class ExportRequest(BaseModel):
    include_optional: bool = True
    branches: list[str] | None = None  # None = all branches


class ExportStats(BaseModel):
    exchange_count: int = 0
    session_count: int = 0
    card_count: int = 0
    branch_count: int = 0
    bookmark_count: int = 0
    annotation_count: int = 0
    variant_count: int = 0


class ImportStats(BaseModel):
    sessions_imported: int = 0
    exchanges_imported: int = 0
    branches_imported: int = 0
    cards_written: int = 0
    trust_modifications_imported: int = 0
    bookmarks_imported: int = 0
    annotations_imported: int = 0
    variants_imported: int = 0
    warnings: list[str] = []


class ImportResponse(BaseModel):
    status: str
    rp_folder: str
    stats: ImportStats


class GuidelinesResponse(BaseModel):
    pov_mode: PovMode | None = None
    dual_characters: list[str] = []
    narrative_voice: NarrativeVoice | None = None
    tense: Tense | None = None
    tone: list[str] | str | None = None
    scene_pacing: ScenePacing | None = None
    integrate_user_narrative: bool | None = None
    pov_character: str | None = None
    preserve_user_details: bool | None = None
    sensitive_themes: list[str] = []
    hard_limits: str | list[str] | None = None
    response_length: ResponseLength | None = None
    include_writing_principles: bool = True
    include_npc_framework: bool = True
    include_output_format: bool = True
    avatar: str | None = None
    body: str | None = None

File: frontend/src/lib/types/rp.ts
export interface RPCreate {
  rp_name: string;
  pov_mode?: 'single' | 'dual';
  dual_characters?: string[];
  tone?: string;
  scene_pacing?: 'slow' | 'moderate' | 'fast';
}
export interface RPResponse {
  rp_folder: string;
  created_files: string[];
}
export interface RPInfo {
  rp_folder: string;
  has_story_cards: boolean;
  card_count: number;
  has_guidelines: boolean;
  has_avatar: boolean;
  branches: string[];
}
export interface GuidelinesResponse {
  pov_mode: 'single' | 'dual' | null;
  dual_characters: string[];
  narrative_voice: 'first' | 'third' | null;
  tense: 'present' | 'past' | null;
  tone: string[] | string | null;
  scene_pacing: 'slow' | 'moderate' | 'fast' | null;
  integrate_user_narrative: boolean | null;
  pov_character: string | null;
  preserve_user_details: boolean | null;
  sensitive_themes: string[];
  hard_limits: string | string[] | null;
  response_length: 'short' | 'medium' | 'long' | null;
  include_writing_principles?: boolean;
  include_npc_framework?: boolean;
  include_output_format?: boolean;
  avatar?: string | null;
  body?: string | null;
}

export interface ChunkingConfig {
  strategy: string;
  chunk_size: number;
  chunk_overlap: number;
}

export interface RechunkResponse {
  status: string;
  total_exchanges: number;
  embedded: number;
  skipped: number;
  failed: number;
}

export interface PromptPreview {
  system_prompt: string;
  sections: string[];
}

export interface ExportRequest {
  include_optional?: boolean;
  branches?: string[] | null;
}

export interface ExportStats {
  exchange_count: number;
  session_count: number;
  card_count: number;
  branch_count: number;
  bookmark_count: number;
  annotation_count: number;
  variant_count: number;
}

export interface ImportStats {
  sessions_imported: number;
  exchanges_imported: number;
  branches_imported: number;
  cards_written: number;
  trust_modifications_imported: number;
  bookmarks_imported: number;
  annotations_imported: number;
  variants_imported: number;
  warnings: string[];
}

export interface ImportResponse {
  status: string;
  rp_folder: string;
  stats: ImportStats;
}
